DFT grid pruning

hokru opened the pull request:
## Description
This PR implements standard grid pruning techniques, where one assumes that the electron density is largely isotropic close to the atomic centre and thus coarse grids are sufficient. Typically the atomic integration sphere is divided into several (3-4) regions where the outer region gets the most accurate grid and the grid get step-wise more coarse towards the inner region(s).

Additionally this PR changes the Bragg-Slater (BS) radii and completes the Treutler-Ahlrichs integration code with the recommended mapping values. Old radii were of unknown origin.

_Because of the updated BS radii, this PR will impact the total energies of any DFT calculation and can change how many grid points are generated **even without pruning**._ 


## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] updated BS radii and extended to 104 elements
- [x] implementation of missing Treutler-Eta mapping parameter
- [x] `DFT_NUCLEAR_SCHEME=SBECKE` ('smooth Becke' after: 10.1063/1.5049435. Better for weak interactions, but no time savings/no screening)
- [x] extends `DFT_PRUNING_SCHEME` keyword with:
  - [x] Treutler grid pruning ``TREUTLER``
  - [x] a more robust grid pruning variant of it ``ROBUST``
- [x] new variables: ``XC GRID TOTAL POINTS``, ``XC GRID SPHERICAL POINTS``, ``XC GRID RADIAL POINTS``
- [x] `print > 2` adds (existing but unused) debug grid print out
- [x] tight DFT tests are updated with new values due to the new Bragg-Slater radii used
- [x] testing and error overview
- [x] option to properly disable all pruning (currently only indirect way)
- [x] tests below indicate a reduction of the grid size by 30 and 45% using `ROBUST` and `TREUTLER` pruning, respectively. This leads to a 15-20% speed up for small molecules.
- [x] the following tests have been updated because of the new BS radii: `dft-b3lyp`
`dft-custom-mgga`, `dft-grad-lr1`, `dft-grad-lr3`, `dft-grad-meta`, `dft-grad1`, `dft-grad2`, `dft-omega`, `dft-psivar`, `dft-vv10`, `dft1`, `dftd3-grad`, `dftd3-version`, `frac-traverse`, `opt13`, `pcmsolver-dft`, `pcmsolver-dipole`
- [x] new test: `dft-pruning`

## Questions
- Pruning enabled by default? -> not this PR, but after further testing yes.

## Checklist
- [x] Documentation
- [x] Pruning Tests added
- [x] All relevant `ctests` pass
(http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge


----
Tests

![grid_osc](https://user-images.githubusercontent.com/1648640/55870801-a6499c80-5b89-11e9-957d-68a08aa8ca41.png)
----
**pytest DFT benchmarks:**

_Interaction energies:_
With the introduced `ROBUST` pruning scheme (less aggressive variant of `TREUTLER`) all `dft_benchmark_interaction` tests pass. With `TREUTLER` 21 of 92 fail.

For comparison, the `FUNCTION` schemes are all very aggressive, failure rate(!): 
- `P_SLATER=39/92`
- `D_SLATER=89/92`
- `LOG_SLATER=61/92`
- `P_GAUSSIAN=92/92` (yes, all fail);
- `D_GAUSSIAN=92/92`
- `LOG_GAUSSIAN=78/92`

_ionization energies:_
- `TREUTLER` 15/92 fail.
- `ROBUST` all pass.

The `FUNCTION` schemes seem still very experimental and are not generally recommendable.

----
**errors in absolute and relative energies**
Based on S22 benchmark set. no CP correction. Absolute energies for dimers only.
level of theory: PBE/aug-cc-pVTZ with grid (434,75)
_grid point reduction_
*  ``ROBUST = 30.3 %``
*  ``TREUTLER = 45.5 %``

_interaction energies_
* MAD old BS to new BS: 0.002 kcal/mol
* MAD old BS to newBS + ``ROBUST``: 0.002 kcal/mol (yay!)
* MAD old BS to newBS + ``TREUTLER``: 0.021 kcal/mol

_absolute energies_
* MAD old BS to new BS: 2.9E-06 au
* MAD old BS to newBS + ``ROBUST``: 2.9E-06 au
* MAD old BS to newBS + ``TREUTLER``: 4.5E-05
* (newBS) MAD ``NONE`` to ``ROBUST`` pruning: 1.4E-10 au
* (newBS) MAD ``NONE`` to ``TREUTLER`` pruning: 4.6E-05 au

----
**timings** (speed ups)
* benzene dimer : `ROBUST=1.16`; `TREUTLER=1.22`
PBE/aug-cc-pVTZ with grid (434,75); 16 Threads; 

* C60 fullerene (C2h) : `ROBUST=1.33`; `TREUTLER=1.36`
PBE/aug-cc-pVDZ with grid (434,75); 16 Threads; 

----
**tough case**
Molecule 35 from the MB16-43 ('mindless benchmarking') benchmark set. Modelling following reaction:
`2x 35 + 20x H2 -> 2x LIH + 4x BH4 + 4x CH4 +  O2 + 4x AlH4 + 2x SiH4 + P2 + CL2`
In essence, a diverse inorganic, main group chemistry reaction.
level of theory: PBE/def2-TZVP with (434,85) grid.
link to input file: https://gist.github.com/hokru/408873d71efb522f7de1f0cfe0cf9da0


_reaction energy_ 
* dev. old BS to new BS: -0.0011 kcal/mol 
* (newBS) `NONE` to `ROBUST`: -0.0003 kcal/mol
* (newBS) `NONE` to `TREUTLER`: 0.94978 kcal/mol (0.5 % of a reaction energy of 188 kcal/mol !)
* speed-ups: `ROBUST=1.33 `; `TREULTER=1.36`

_absolute energy molecule 35_
* dev. old BS to new BS: 3.41e-07
* (newBS) `NONE` to `ROBUST`:  4.93e-07
* (newBS) `NONE` to `TREUTLER`: 1.46E-04

_gradient energy molecule 35_
* (newBS) MAD `NONE` to `ROBUST`: ~ 1E-7
* (newBS) MAD `NONE` to `TREUTLER`: ~ 1E-05

_grid point reduction molecue 35_
*  ``ROBUST =  30.9 %``
*  ``TREUTLER = 45.8 %``

dgasmith commented:
Cool! Glad someone is looking into the grid code a bit. Traveling a bit now, but will think on your questions.

susilehtola commented on psi4/src/psi4/libfock/cubature.cc:
```diff
@@ -56,9 +56,78 @@ using namespace psi;
+      2.929, 2.740, 2.551, 2.551, 2.457, 2.551, 2.551, 2.551, 2.835, 3.591, 3.024, 3.024, 3.591, 3.591, 3.591, // Rn
+      4.063, 4.063,                                                                //Fr-Ra
+             3.685, 3.401,3.401,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,//Ac-Lr
+    };
```
These should be ```std::vector<double>```

CDSherrill commented:
Holger, very glad you are taking this on!

On Fri, Apr 5, 2019, 2:32 AM Holger Kruse <notifications@github.com> wrote:

> Description
>
> This *work-in-progress* PR implements standard grid pruning techniques,
> where one assumes that the electron density is largely isotropic close to
> the atomic centre and thus coarse grids are sufficient. Typically the
> atomic integration sphere is divided into several (3-4) regions where the
> outer region gets the most accurate grid and the grid get step-wise more
> coarse towards the inner region(s).
>
> Additionally this PR changes the Bragg-Slater (BS) radii and completes the
> Treutler-Ahlrichs integration code with the recommended mapping values. Old
> radii were of unknown origin.
>
> *Because of the updated BS radii, this PR will impact the total energies
> of any DFT calculation and can change how many grid points are generated
> even without pruning.* (11/33 DFT tests fail)
> Todos
>
> Notable points (developer or user-interest) that this PR has or will
> accomplish.
>
>    - updated BS radii and extended to 104 elements
>    - implementation of missing Treutler-Eta mapping parameter
>    - DFT_PRUNE_TYPE keyword:
>       - Treutler grid pruning (DFT_PRUNE_TYPE=TREUTLER)
>       - a more robust grid pruning (DFT_PRUNE_TYPE=ROBUST)
>       - old pruning using a function (DFT_PRUNE_TYPE=FUNCTION)
>    - modified Becke mu from Ochsenfeld for smoother integration (however
>    no time savings)
>    - possibility to scale Treutler-Etas to better treat diffuse basis sets
>    - better print out of options
>    - testing and error overview
>    - option to properly disable all pruning (currently only indirect way)
>
> Questions
>
>    - How to C++? Explicit guidance or help is welcome.
>    - All the very tight DFT and VV10 tests break or will break. How to
>    'solve' that?
>    - Do we want to be able to switch between old and new BS radii (as
>    expert-only option)?
>    - DFT_PRUNING_SCHEME re-work to not have DFT_PRUNE_TYPE as competing
>    and confusing option?
>    - Do users want access to custom pruning, e.g. via an array to define
>    grids for RadialPruneMgr::ShellPruning ?
>    - Pruning enabled by default? (I would vote yes)
>
> Checklist
>
>    - Documentation
>    - Tests added
>    - All or relevant fraction of full tests run
>    <http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests>
>
> Status
>
>    - Ready for review
>    - Ready for merge
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/psi4/psi4/pull/1591
> Commit Summary
>
>    - first try of Treutler pruning. Error very large
>    - testing
>    - testing stage
>    - adds Treutlers M4 fitting radii and new BS radii
>    - adds switch for 3 different pruning schemes
>    - BSrad in Angstrom works best..
>    - cleaning up
>    - adds smooth Becke mu of Ochsenfeld
>
> File Changes
>
>    - *M* psi4/src/psi4/libfock/cubature.cc
>    <https://github.com/psi4/psi4/pull/1591/files#diff-0> (235)
>    - *M* psi4/src/psi4/libfock/cubature.h
>    <https://github.com/psi4/psi4/pull/1591/files#diff-1> (9)
>    - *M* psi4/src/read_options.cc
>    <https://github.com/psi4/psi4/pull/1591/files#diff-2> (5)
>    - *M* tests/dft-b3lyp/input.dat
>    <https://github.com/psi4/psi4/pull/1591/files#diff-3> (3)
>
> Patch Links:
>
>    - https://github.com/psi4/psi4/pull/1591.patch
>    - https://github.com/psi4/psi4/pull/1591.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/psi4/psi4/pull/1591>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC9QdvDQ8k3q9iMmt3WyUH9_hci95FKbks5vdu4LgaJpZM4ceQFR>
> .
>

hokru commented:
added plot for Argon dimer showing oscillating energy (following the Ochsenfeld paper)

andysim commented:
That red line is wild!  Thanks for the plot - it's really helpful.

susilehtola commented:
Would you consider implementing parameter-free radial grids on the same go? See e.g. [10.1063/1.1759323](http://dx.doi.org/10.1063/1.1759323)

hokru commented:
> Would you consider implementing parameter-free radial grids on the same go? See e.g. [10.1063/1.1759323](http://dx.doi.org/10.1063/1.1759323)

Hmm, unlikely. I have limited time for coding and it needs to directly impact my science, either by greatly speeding things up or by adding to a paper. I could imagine doing a psi4-specific iterative parameter-free grid generation in the same spirit, but not directly their scheme.

hokru commented:
I am looking to update the failing dft test where absolute energies are tested. Looking for opinions and advice:

options:
  * A) reduce the comparison thresholds, where necessary. Down to 5 or 6  (from 7 or even higher)
  * B) make new reference values using the new BS radii. Perhaps cleanest test history about the origin of the numbers and best for gradients
  * C) a mixture of both.

these `ctest -L dft` fail on my machine:
        157 - dft-grad-lr1 (Failed)
        159 - dft-grad-lr3 (Failed)
        167 - dft-grad-meta (Failed)
        169 - dft-grad1 (Failed)
        170 - dft-grad2 (Failed)
        171 - dft-psivar (Failed)
        173 - dft1 (Failed)
        174 - dft-vv10 (Failed)
        178 - dft-omega (Failed)
        390 - dft-custom-mgga (Failed)
        423 - dftd3-grad (Failed)
        431 - pcmsolver-dft (Failed)

loriab commented:
I think (B) with a comment in the tests about why updated. We want to keep checking to the same precision, but like physconst updates, the ref value has changed.

PeterKraus commented:
(B) seconded. All of the tests are total energies / gradients, so I think it's fine to update the reference values, as long as the shifts aren't huge. 

Related, the `dft-custom-mgga` could be retired in favour of the interaction/ionisation tests, or is it covering something that those don't?

hokru commented:
> Related, the `dft-custom-mgga` could be retired in favour of the interaction/ionisation tests, or is it covering something that those don't?

It, and the others like it, tests absolute energies instead of relative ones.

PeterKraus commented:
> > Related, the `dft-custom-mgga` could be retired in favour of the interaction/ionisation tests, or is it covering something that those don't?
> 
> It, and the others like it, tests absolute energies instead of relative ones.

True. Could be moved to `pytests` though, so that the testing of DFT functionals is done in one place, with the exception of smoke tests...

loriab commented:
Agree with not dropping tests (esp. when have other programs' numbers) but fine to rearrange them, especially when `pytest.mark.parametrize` could make them more readable. I expect the `dft-*-custom`s could be consolidated. But, yes, beyond scope of this PR.

hokru commented:
Proposal for keyword reorganization, replacing `DFT_PRUNING_SCHEME` like so:

```
DFT_PRUNING_SCHEME  = 'FUNCTION' 'REGION'
DFT_PRUNING_FUNCTION = 'FLAT' 'P-SLATER' '...' 
DFT_PRUNING_REGION = 'ROBUST' 'TREUTLER'
```

Maybe current `DFT_PRUNING_SCHEME` could handle both sub-options somehow? but I don't see an easy way to do it.

hokru commented:
To explain. The current `DFT_PRUNING_SCHEME` selects a pruning-function (flat, slater, gaussian type). This PR added `DFT_PRUNING_TYPE` (see PR description), but _SCHEME vs _TYPE is confusing, imo. Each approach has sub-options.

dgasmith commented:
I am not aware of an area that we have made branching options like this before. @loriab may have a location. Barring any prior art I think the current layout makes plenty of sense.

Apologies for being MIA a bit on this PR. I am **very** happy that someone is looking into the grid files! All of the changes look good at cursory glance. Please do ping me if something specific comes up, I think these kinds of changes are very important to the future of the DFT code.

hokru commented:
I changed the keywords following my proposal. Also will remove the WIP in the topic as this is getting closer to actually being finished.

loriab commented:
Apologies if I didn't trace the logic thoroughly and correctly. But doesn't one always do region _or_ function (or flat)? From the user's perspective, could one then do

```
/*- Select approach for pruning. Options ``ROBUST`` and ``TREUTLER`` prune based on regions (proximity to nucleus) while ... !expert -*/
options.add_str("DFT_PRUNING_TYPE", "ROBUST", "FLAT P_GAUSSIAN D_GAUSSIAN P_SLATER D_SLATER LOG_GAUSSIAN LOG_SLATER ROBUST TREUTLER");
```

I think the most related option was https://github.com/psi4/psi4/blob/master/psi4/src/read_options.cc#L177 where it could have been two options but we rolled into one for the user.

hokru commented:
A single keyword would perhaps be best for the user. I will try my luck with that.

hokru commented:
Changed it back to a single ``DFT_PRUNING_SCHEME`` option and added a basic pruning test.
The new ``DFT_NUCLEAR_SCHEME=SBECKE`` has no test, but the nuclear and radial building schemes have no tests in general...

hokru commented:
*Question*: Do we want to be able to switch between old and new BS radii (as expert option)?

Personally I think I would leave them commented out in the code and not make a user-facing switch available.

dgasmith commented:
I think that should be sufficient, the differences are nominal enough. On this note can you quantify the energy differences (both absolute and relative) so that we can mention this in the release notes?

hokru commented:
I added various error statistics over the S22 benchmark set.

hokru commented:
I want to go through the code again and make sure everything is clean before this can be thoroughly reviewed.

dgasmith commented:
Very impressive grid reduction for a tiny impact! This is really awesome!

hokru commented:
I suppose this is ready for review!  :-)

dgasmith commented on doc/sphinxman/source/dft.rst:
```diff
@@ -531,16 +530,18 @@ octohedral, and icosohedral systems are not complete, so there may be some
+
+Pruning of the quadrature grid is controlled by the |scf__dft_pruning_scheme|
+option. The options ``ROBUST`` and ``TREUTLER`` devide the grid into spherical regions
+based on the Bragg-Slater radius of each atom and apply different orders to them. The ``ROBUST`` scheme is a less aggressive variant of the ``TREUTLER`` approach and suitable for benchmark-level quality (MAD < 0.002 kcal/mol for the S22 with PBE/aug-cc-pVTZ for pruned versus unpruned grid). Our implementation of the ``TREUTLER`` scheme shows an error of 0.02 kcal/mol for the same bechmark. Both a ...
```
This needs newlines like above.

We should add a note here that the pruning option was introduced in `v1.4` and may change the energies compared to previous versions by the aforementioned 0.002 kcal/mol relative or 1.e-6 absolute.

@cdsgroup we may need to recruit someone to do a deeper investigation here.

hokru replied:
added a note on top.

dgasmith commented on psi4/src/psi4/libfock/cubature.cc:
```diff
@@ -2808,14 +2848,16 @@ double NuclearWeightMgr::computeNuclearWeight(MassPoint mp, int A, double stratm
     for (int l = 0; l < natom; l++) dist[l] = distToAtom(mp, l);
 
     double (*stepFunction)(double) = (scheme_ == STRATMANN) ? StratmannStepFunction : BeckeStepFunction;
+    double (*muFunction)(double,double,double) = (scheme_ == SBECKE) ? SmoothBeckeMu : BeckeMu;
```
Can you run clang-format over these files?

hokru replied:
done

hokru commented:
I will add some timings and check the gradient soon(TM).
The various dft gradient tests do generally _not_ pass with `ROBUST` pruning. Didn't check errors, but those tests are often very tight.

Finally, pruning _on_ by default (`ROBUST` scheme)?? I'd argue that is what users expect since most (all?) big QM programs do this. Possibly wants more thorough testing.

dgasmith commented:
I agree that `ROBUST` should be the default. We might want to leave the default to no pruning and then make an issue reviewing the timings, accuracy, and change log notes before switching the default over. Whenever we change the results the DFT code obtains even if minor I do get a little nervous.

hokru commented:
added some timings and a gradient error to the post.

hokru commented on doc/sphinxman/source/dft.rst:
```diff
@@ -531,16 +536,27 @@ octohedral, and icosohedral systems are not complete, so there may be some
 The atomic weighting scheme is controlled by the |scf__dft_nuclear_scheme|
-option, which may be one of TREUTLER, BECKE, or NAIVE.
+option, which may be one of ``TREUTLER``, ``BECKE``, ``STRATMANN``, ``NAIVE``,
+or ``SBECKE``. The latter is a smoother variant of the BECKE scheme recently introduced 
```
fixed

hokru commented on doc/sphinxman/source/dft.rst:
```diff
@@ -531,16 +536,27 @@ octohedral, and icosohedral systems are not complete, so there may be some
+by Laqua [Laqua:2018:204111]_ that should behave better for weak interactions.
+
+Pruning of the quadrature grid is controlled by the |scf__dft_pruning_scheme|
+option. The options ``ROBUST`` and ``TREUTLER`` devide the grid into spherical regions
```
fixed

hokru commented on doc/sphinxman/source/dft.rst:
```diff
@@ -531,16 +536,27 @@ octohedral, and icosohedral systems are not complete, so there may be some
+The ``ROBUST`` scheme is a less aggressive variant of the ``TREUTLER`` approach and
+suitable for benchmark-level quality (MAD < 0.002 kcal/mol for the S22 with PBE/aug-cc-pVTZ
+for pruned versus unpruned grid). Our implementation of the ``TREUTLER`` scheme shows an 
+error of 0.02 kcal/mol for the same bechmark. Both also reduce the grid order by 1 for H 
```
fixed

hokru commented on psi4/src/psi4/libfock/cubature.cc:
```diff
@@ -3497,6 +3543,71 @@ int RadialPruneMgr::GetPrunedNumAngPts(double rho) {
+    // H, He always 1 smaller
+    if (Z <= 2) {
+        pruned_order = nominal_order_ - 1;
+    };  // ToDo: flag to enable/disable
```
ok!

hokru commented on psi4/src/psi4/libfock/cubature.cc:
```diff
@@ -3456,38 +3500,40 @@ class RadialPruneMgr {
+    for (size_t i = 0; i < sizeof(prunefunctions) / sizeof(prunefunctions[0]); i++)
+        if (strcmp(prunefunctions[i].name, functionname) == 0) return i;
+    outfile->Printf("Unrecognized pruning function name  %s!\n", functionname);
+    throw PSIEXCEPTION("Unrecognized pruning function name!");
```
maybe?
All 4 `WhichScheme`s functions use `chars`, historically I assume. I only renamed a variable and did not write this.

Changing all `char` to `std::string` does not seem to work. I still get multiple error char-to-string errors. Not sure what is going on.
```
/Users/kruse/qc/psi4/psi4/src/psi4/libfock/cubature.cc:3525:90: error: cannot convert ‘const char*’ to ‘const string*’ {aka ‘const std::__cxx11::basic_string<char>*’} in initialization
                                                                        {nullptr, nullptr}};
                                                                                          ^
```

loriab replied:
Got it. Looking closer, this was written in the `std::string` era, so I'll suppose the authors used `char` for good reason.

hokru commented on psi4/src/psi4/libfock/cubature.cc:
```diff
@@ -3497,6 +3543,71 @@ int RadialPruneMgr::GetPrunedNumAngPts(double rho) {
+    // S22 interaction energy OK with above
+    if (ri <= nr) {  // <= 25
+        pruned_order = region4;
+    } else if (ri > nr && ri <= 2 * nr) {  // > 25 && <= 50
```
hmm yes, seems you are right.

dgasmith commented on psi4/src/psi4/libfock/cubature.cc:
```diff
@@ -56,24 +56,46 @@ using namespace psi;
+             3.685, 3.496,3.496,3.496,3.496,3.496,3.496,3.402,3.307,3.307,3.307,3.307,3.307,3.307,3.307,//*La-Lu
+      2.929, 2.740, 2.551, 2.551, 2.457, 2.551, 2.551, 2.551, 2.835, 3.591, 3.024, 3.024, 3.591, 3.591, 3.591, // Rn
+      4.063, 4.063,                                                                //Fr-Ra
+             3.685, 3.401,3.401,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,3.307,//Ac-Lr
```
Makes sense.

hokru commented:
Added changes requested by lori and tried to fix the LGTM report as well. Found a bug in the SBECKE scheme through it.

loriab commented on tests/dft-pruning/input.dat:
```diff
@@ -0,0 +1,31 @@
+  set DFT_PRUNING_SCHEME $scheme
+  energy('pbe/sto-3g')
+  points=variable('XC GRID TOTAL POINTS')
+  compare_values(SCHEMES[scheme],points,2,'grid points for '+scheme) #TEST
```
since these don't need floats, can do `compare_integers(SCHEMES[scheme], int(variable('XC GRID TOTAL POINTS')), 'grid porints for '+scheme)`

loriab commented on tests/dft-grad2/input.dat:
```diff
@@ -1,9 +1,10 @@
-    [-0.02563442179393,  0.01685678274170, 0.00000000000000]        #TEST
+    [ 0.015033129924,    -0.012524858341,     0.000000000000],       #TEST
+    [ 0.010601799255,    -0.004332928749,     0.000000000000],       #TEST
+    [-0.025634252349,     0.016856185932,     0.000000000000]        #TEST
```
Did you generate the new ref vals with tight convcrit? There's a few guidelines in the last paragraph of [here](http://psicode.org/psi4manual/master/add_tests.html). Or ask for any advice. Otherwise problems can arise on different computers just due to variation in number of iterations, etc.

hokru replied:
no, i did not. Will have to change that. I used the settings from the tests themselves. I assume this is mainly an issue of the gradient tests?

loriab replied:
It hits energy tests, too, unfortunately. set e_/d_/r_convergence to 10, collect the ref numbers, remove the convergences, and tests should be good to go.

hokru replied:
updated the tests with tighter convergence.

loriab commented on psi4/src/read_options.cc:
```diff
@@ -1485,16 +1485,18 @@ int read_options(const std::string &name, Options &options, bool suppress_printi
-                        "FLAT P_GAUSSIAN D_GAUSSIAN P_SLATER D_SLATER LOG_GAUSSIAN LOG_SLATER");
+        /*- Select approach for pruning. Options ``ROBUST`` and ``TREUTLER`` prune based on regions (proximity to nucleus) while
+        ``FLAT`` ``P_GAUSSIAN`` ``D_GAUSSIAN`` ``P_SLATER`` ``D_SLATER`` ``LOG_GAUSSIAN`` ``LOG_SLATER`` prune based on decaying functions (experts only!).
+        The recommended option are ``ROBUST`` or ``TREUTLER``. -*/
```
Treutler's still pretty risky, isn't it? Aren't the recommended options "ROBUST" or "NONE"?

hokru replied:
I will change it to "The recommended scheme is ``ROBUST``", if that's alright.

loriab commented on psi4/src/psi4/libfock/cubature.cc:
```diff
@@ -56,24 +56,46 @@ using namespace psi;
-        1.485, 1.474, 1.562, 1.562, 1.562, 1.562, 1.562, 1.562, 1.562, 1.562, 1.562, 1.562, 1.650, 1.727, 1.760, 1.771, 1.749, 1.727,
-        1.628, 1.606, 1.639, 1.639, 1.639, 1.639, 1.639, 1.639, 1.639, 1.639, 1.639, 1.639, 1.672, 1.804, 1.881, 1.892, 1.892, 1.881,
+// clang-format off
+double GetBSRadiusNew(unsigned Z) {
```
feel free to name this `GetBSRadius`, so its calling doesn't change elsewhere.
